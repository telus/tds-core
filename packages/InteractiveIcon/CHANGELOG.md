# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [2.0.2](https://github.com/telus/tds-core/compare/@tds/core-interactive-icon@2.0.1...@tds/core-interactive-icon@2.0.2) (2020-07-22)

**Note:** Version bump only for package @tds/core-interactive-icon





## [2.0.1](https://github.com/telus/tds-core/compare/@tds/core-interactive-icon@2.0.0...@tds/core-interactive-icon@2.0.1) (2020-04-14)

**Note:** Version bump only for package @tds/core-interactive-icon





# [2.0.0](https://github.com/telus/tds-core/compare/@tds/core-interactive-icon@1.4.7...@tds/core-interactive-icon@2.0.0) (2020-03-17)


### Features

* **core-interactive-icon:** add color prop to dependent icon ([ef70f1e](https://github.com/telus/tds-core/commit/ef70f1e29553cf5dafb8b6d4eb7390211a39db43))
* **core-interactive-icon:** replace InteractiveIcon with IconButton ([b29f2f5](https://github.com/telus/tds-core/commit/b29f2f5abd0918544c75dc8964c138fff47e53ab))


### BREAKING CHANGES

* **core-interactive-icon:** move all InteractiveIcon components to Dependent, add IconButton wrapper component





## [1.4.7](https://github.com/telus/tds-core/compare/@tds/core-interactive-icon@1.4.6...@tds/core-interactive-icon@1.4.7) (2020-03-11)

**Note:** Version bump only for package @tds/core-interactive-icon





## [1.4.6](https://github.com/telus/tds-core/compare/@tds/core-interactive-icon@1.4.5...@tds/core-interactive-icon@1.4.6) (2020-03-10)

**Note:** Version bump only for package @tds/core-interactive-icon





## [1.4.5](https://github.com/telus/tds-core/compare/@tds/core-interactive-icon@1.4.4...@tds/core-interactive-icon@1.4.5) (2020-03-09)

**Note:** Version bump only for package @tds/core-interactive-icon





## [1.4.4](https://github.com/telus/tds-core/compare/@tds/core-interactive-icon@1.4.3...@tds/core-interactive-icon@1.4.4) (2020-02-12)


### Bug Fixes

* **core-interactive-icon:** prevent props are undefined navbutton errors ([e90435d](https://github.com/telus/tds-core/commit/e90435d))





## [1.4.3](https://github.com/telus/tds-core/compare/@tds/core-interactive-icon@1.4.2...@tds/core-interactive-icon@1.4.3) (2020-02-03)

**Note:** Version bump only for package @tds/core-interactive-icon





## [1.4.2](https://github.com/telus/tds-core/compare/@tds/core-interactive-icon@1.4.1...@tds/core-interactive-icon@1.4.2) (2020-01-27)

**Note:** Version bump only for package @tds/core-interactive-icon





## [1.4.1](https://github.com/telus/tds-core/compare/@tds/core-interactive-icon@1.4.0...@tds/core-interactive-icon@1.4.1) (2020-01-23)

**Note:** Version bump only for package @tds/core-interactive-icon





# [1.4.0](https://github.com/telus/tds-core/compare/@tds/core-interactive-icon@1.3.2...@tds/core-interactive-icon@1.4.0) (2020-01-22)


### Features

* **core-interactive-icon:** add dependent icon ([e9c90e1](https://github.com/telus/tds-core/commit/e9c90e1))





## [1.3.2](https://github.com/telus/tds-core/compare/@tds/core-interactive-icon@1.3.1...@tds/core-interactive-icon@1.3.2) (2020-01-07)


### Bug Fixes

* **core-interactive-icon:** change invert cart filled dot to telus green ([3fd46cc](https://github.com/telus/tds-core/commit/3fd46cc))
* **core-interactive-icon:** improve invert navbutton focus/active style ([4064f0b](https://github.com/telus/tds-core/commit/4064f0b))





## [1.3.1](https://github.com/telus/tds-core/compare/@tds/core-interactive-icon@1.3.0...@tds/core-interactive-icon@1.3.1) (2019-11-18)

**Note:** Version bump only for package @tds/core-interactive-icon





# [1.3.0](https://github.com/telus/tds-core/compare/@tds/core-interactive-icon@1.2.1...@tds/core-interactive-icon@1.3.0) (2019-11-06)


### Features

* **core-interactive-icon:** add NavButton icons ([a511499](https://github.com/telus/tds-core/commit/a511499))





## [1.2.1](https://github.com/telus/tds-core/compare/@tds/core-interactive-icon@1.2.0...@tds/core-interactive-icon@1.2.1) (2019-10-09)

**Note:** Version bump only for package @tds/core-interactive-icon





# [1.2.0](https://github.com/telus/tds-core/compare/@tds/core-interactive-icon@1.1.0...@tds/core-interactive-icon@1.2.0) (2019-10-04)


### Features

* **core-interactive-icon:** add caret and chevron icons ([eaa6bc2](https://github.com/telus/tds-core/commit/eaa6bc2))





# [1.1.0](https://github.com/telus/tds-core/compare/@tds/core-interactive-icon@1.0.2...@tds/core-interactive-icon@1.1.0) (2019-09-18)


### Features

* **core-interactive-icon:** allow forwarding refs ([a9dbe73](https://github.com/telus/tds-core/commit/a9dbe73))





## [1.0.2](https://github.com/telus/tds-core/compare/@tds/core-interactive-icon@1.0.1...@tds/core-interactive-icon@1.0.2) (2019-09-12)

**Note:** Version bump only for package @tds/core-interactive-icon





## [1.0.1](https://github.com/telus/tds-core/compare/@tds/core-interactive-icon@1.0.0...@tds/core-interactive-icon@1.0.1) (2019-09-06)

**Note:** Version bump only for package @tds/core-interactive-icon





# 1.0.0 (2019-08-26)


### Features

* **core-interactive-icon:** add InteractiveIcon component ([f882229](https://github.com/telus/tds-core/commit/f882229))
