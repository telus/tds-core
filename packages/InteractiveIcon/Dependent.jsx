import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'

import { Consumer } from '../../shared/utils/context'
import { warn } from '../../shared/utils/warn'

const positionStyles = ({ paragraphSize }) => {
  let top = 0
  if (paragraphSize === 'large') {
    top = '-4px'
  }

  return {
    position: 'relative',
    top,
  }
}

const invertStyles = ({ invert }) => {
  return {
    fill: invert && '#FFFFFF',
  }
}
export const StyledDependentSVG = styled.svg(positionStyles, invertStyles, ({ paragraphSize }) => ({
  width: paragraphSize === 'small' ? 20 : 24,
  height: paragraphSize === 'small' ? 20 : 24,
}))

const Dependent = ({ invert, children, parentType, ...rest }) => {
  if (parentType !== 'a' && parentType !== 'button') {
    warn('Dependent', 'Icon must be contained in an <a> or <button>')
  }

  return (
    <Consumer>
      {({ paragraphSize }) => {
        return React.cloneElement(children, { paragraphSize, invert, ...rest })
      }}
    </Consumer>
  )
}

Dependent.propTypes = {
  /**
   * @ignore
   */
  parentType: PropTypes.oneOf(['a', 'button']),
  /**
   * @ignore
   */
  invert: PropTypes.bool,
  /**
   * @ignore
   */
  children: PropTypes.node.isRequired,
}

Dependent.defaultProps = {
  parentType: undefined,
  invert: false,
}

export default Dependent
