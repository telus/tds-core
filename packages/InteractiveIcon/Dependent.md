### Usage criteria

These icons indicate interactivity, but are not interactive on their own. They depend on the [Link](#/Links?id=link) component.

- Must include accessible hidden text using `A11yContent` when label does not explicitly indicate use or action
  - Be mindful of using `A11yContent` appropriately as it may appear before or after the label
- Must not be used in headings
- Must not be used as an independent interactive component or render as independent SVG
- Should depend on other interactive components with a visible label as part of the wrapper component
- Dependent icons are 20x20 when used next to small (14px) links, or 24x24 when used next to medium or large links

```jsx noeditor
const IconTable = require('../../docs/components/custom/IconTable/IconTable').default
;<Box between={4}>
  <IconTable
    heading="Icons for use with Links"
    icons={[
      {
        name: 'LinkExternal',
        Component: LinkExternal,
        usageCriteria:
          'Indicate a reference to an external source; will direct away from the TELUS experience or the primary flow ',
      },
    ]}
  />
  <IconTable
    heading="Icons for use with Buttons"
    icons={[
      {
        name: 'Add',
        Component: Add,
        usageCriteria: 'Used to add items',
        props: {
          a11yText: 'Add',
        },
      },
      {
        name: 'Close',
        Component: Close,
        usageCriteria: 'Used to close',
        props: {
          a11yText: 'Close',
        },
      },
      {
        name: 'PlayVideo',
        Component: PlayVideo,
        usageCriteria:
          'Used to indicate that a video is available, used with text link to bring to another page or window',
        props: {
          a11yText: 'Play',
        },
      },
      {
        name: 'Subtract',
        Component: Subtract,
        usageCriteria: 'Used to reduce items',
        props: {
          a11yText: 'Remove',
        },
      },
      {
        name: 'Delete',
        Component: Delete,
        usageCriteria:
          'Delete or edit an item. Use for authenticated or internal applications. Don’t use on marketing pages.',
        props: {
          a11yText: 'Delete',
        },
      },
    ]}
  />
  <IconTable
    heading="Icons for use with Links or Buttons"
    icons={[
      {
        name: 'Download',
        Component: Download,
        usageCriteria: 'Indicate download function',
      },
      {
        name: 'DownloadPDF',
        Component: DownloadPDF,
        usageCriteria: 'Indicate download of a single PDF',
      },
      {
        name: 'DownloadPDFs',
        Component: DownloadPDFs,
        usageCriteria: 'Indicate download of multiple PDFs',
      },
      {
        name: 'Edit',
        Component: Edit,
        usageCriteria:
          'Indicate ability to edit. Use for authenticated or internal applications. Don’t use on marketing pages',
      },
      {
        name: 'Print',
        Component: Print,
        usageCriteria: 'Prompt native print module',
      },
      {
        name: 'Profile',
        Component: Profile,
        usageCriteria: 'Direct to, or indicate user profile',
      },
      {
        name: 'Search',
        Component: Search,
        usageCriteria: 'Direct to, or indicate search functions',
      },
      {
        name: 'Settings',
        Component: Settings,
        usageCriteria: 'Direct to, or indicate available settings',
      },
    ]}
  />
</Box>
```

### Minimal usage

```jsx
<Link href="#" icon={Print} iconPosition="left">
  Print
</Link>
```

### Inverted

To invert a `DependentIcon`, set invert on the parent `Link`

```jsx
<Box inset={3} style={{ background: '#4b286d' }}>
  <Link href="#" icon={Print} iconPosition="left" invert>
    Print
  </Link>
</Box>
```
