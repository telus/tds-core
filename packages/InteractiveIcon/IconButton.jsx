import React, { forwardRef } from 'react'
import PropTypes from 'prop-types'
import styled, { ThemeProvider } from 'styled-components'

import A11yContent from '@tds/core-a11y-content'
import { colorGreyGainsboro } from '@tds/core-colours'
import { safeRest } from '@tds/util-helpers'
import { componentWithName } from '@tds/util-prop-types'

import animations from './shared/animations'
import SharedStyledInteractiveIconButton from './shared/StyledInteractiveIconButton'
import SharedStyledInteractiveIconHover from './shared/StyledInteractiveIconHover'
import { warn } from '../../shared/utils/warn'

export const StyledIconButton = styled(SharedStyledInteractiveIconButton)(
  animations.scale,
  animations.reduceMotion
)

const getTheme = variant => {
  if (variant === 'alternative') {
    return {
      hoverBackgroundColor: '#D8CBE5',
    }
  }
  if (variant === 'inverted') {
    return {
      hoverBackgroundColor: 'transparent',
    }
  }
  return {
    hoverBackgroundColor: colorGreyGainsboro,
  }
}

/**
 * @version ./package.json
 */
const IconButton = forwardRef(({ a11yText, variant, onClick, tag, icon: Icon, ...rest }, ref) => {
  let color
  if (variant === 'alternative') {
    color = 'telusPurple'
  } else if (variant === 'inverted') {
    color = 'white'
  } else {
    color = 'greyShark'
  }

  if (
    Icon.name !== 'Add' &&
    Icon.name !== 'Close' &&
    Icon.name !== 'Subtract' &&
    Icon.name !== 'PlayVideo'
  ) {
    warn(
      'IconButton',
      'IconButton is meant to be used with the Add, Close, Subtract, and PlayVideo icons for their universally-recognizable appearance. Other icons should be accompanied with text and not as a part of IconButton.'
    )
  }

  return (
    <ThemeProvider theme={getTheme(variant)}>
      <StyledIconButton {...safeRest(rest)} variant={variant} onClick={onClick} as={tag} ref={ref}>
        <A11yContent>{a11yText}</A11yContent>
        <SharedStyledInteractiveIconHover />
        <Icon color={color} />
      </StyledIconButton>
    </ThemeProvider>
  )
})

IconButton.displayName = 'IconButton'

IconButton.propTypes = {
  /**
   * A description of the icon for assistive technology.
   */
  a11yText: PropTypes.string.isRequired,
  /**
   * The visual style.
   */
  variant: PropTypes.oneOf(['default', 'alternative', 'inverted']),
  /**
   * Pass a handler to the icon to make it interactive.
   */
  onClick: PropTypes.func,
  /**
   * The tag
   */
  tag: PropTypes.oneOf(['button', 'a']),
  /**
   * The icon. Can be one of either the `Add`, `Close`, `PlayVideo`, or `Subtract` icons.
   */
  icon: PropTypes.oneOfType([
    componentWithName('Add'),
    componentWithName('Close'),
    componentWithName('PlayVideo'),
    componentWithName('Subtract'),
  ]).isRequired,
}

IconButton.defaultProps = {
  variant: 'default',
  onClick: undefined,
  tag: 'button',
}

export default IconButton
