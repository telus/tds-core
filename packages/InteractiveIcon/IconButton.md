`IconButton` can be used to contain universally-recognizable icons that do not require accompanying text. It can be used with the `Add`, `Close`, `PlayVideo`, and `Subtract` [Dependent icons](#/Icons?id=dependent).

```js
initialState = {
  count: 0,
}

const addCount = () => {
  setState({ count: state.count + 1 })
}
const reduceCount = () => {
  setState({ count: state.count - 1 })
}
;<Box inline between={3} style={{ alignItems: 'center' }}>
  <IconButton icon={Subtract} onClick={reduceCount} a11yText="Reduce" />
  <Text size="medium">{state.count}</Text>
  <IconButton icon={Add} onClick={addCount} a11yText="Add" />
</Box>
```
