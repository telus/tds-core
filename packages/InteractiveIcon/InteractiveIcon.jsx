import React from 'react'
import PropTypes from 'prop-types'
import styled, { ThemeProvider } from 'styled-components'

import A11yContent from '@tds/core-a11y-content'
import { colorGainsboro, colorShark, colorTelusPurple, colorWhite } from '@tds/core-colours'
import { buttons } from '@tds/shared-styles'

import safeRest from '../../shared/utils/safeRest'

const StyledInteractiveIconButton = styled.button(buttons.noStyle, {
  width: '40px',
  height: '40px',
  display: 'inline-flex',
  justifyContent: 'center',
  alignItems: 'center',
  position: 'relative',
  '&:focus': {
    outline: 'none',
  },
})

export const StyledInteractiveIconSVG = styled.svg(({ theme }) => ({ fill: theme.iconColor }), {
  width: '24px',
  height: '24px',
  zIndex: '2',
  transition: 'transform 150ms ease-in-out',
  '@media (prefers-reduced-motion: reduce)': {
    transition: 'none',
  },
  '&:hover': {
    transform: 'scale(1.1, 1.1)',
  },
  '&:active': {
    transform: 'scale(1, 1)',
  },
})

const StyledInteractiveIconHover = styled.span(
  ({ theme }) => ({ backgroundColor: theme.backgroundColor }),
  {
    position: 'absolute',
    top: 0,
    left: 0,
    width: '100%',
    height: '100%',
    zIndex: '1',
    borderRadius: '50%',
    transition: 'transform 200ms ease-in-out',
    '@media (prefers-reduced-motion: reduce)': {
      transition: 'none',
    },
    transform: 'scale(0,0)',
    [`${StyledInteractiveIconButton}:focus &, ${StyledInteractiveIconButton}:active &`]: {
      transform: 'scale(1,1)',
    },
  }
)

const getTheme = variant => {
  if (variant === 'alternative') {
    return {
      backgroundColor: '#D8CBE5',
      iconColor: colorTelusPurple,
    }
  }
  if (variant === 'inverted') {
    return {
      backgroundColor: 'transparent',
      iconColor: colorWhite,
    }
  }
  return {
    backgroundColor: colorGainsboro,
    iconColor: colorShark,
  }
}

/**
 * @version ./package.json
 */
const InteractiveIcon = ({ a11yText, variant, onClick, children, tag, ...rest }) => (
  <ThemeProvider theme={getTheme(variant)}>
    <StyledInteractiveIconButton {...safeRest(rest)} onClick={onClick} as={tag}>
      <A11yContent>{a11yText}</A11yContent>
      <StyledInteractiveIconHover />
      {children}
    </StyledInteractiveIconButton>
  </ThemeProvider>
)

InteractiveIcon.propTypes = {
  /**
   * A description of the icon for screen readers.
   */
  a11yText: PropTypes.string.isRequired,
  /**
   * The style.
   */
  variant: PropTypes.oneOf(['default', 'alternative', 'inverted']),
  /**
   * Pass a handler to the icon to make it interactive.
   */
  onClick: PropTypes.func,
  /**
   * The tag
   */
  tag: PropTypes.oneOf(['button', 'a']),
  /**
   * @ignore
   */
  children: PropTypes.node.isRequired,
}

InteractiveIcon.defaultProps = {
  variant: 'default',
  onClick: undefined,
  tag: 'button',
}

export default InteractiveIcon
