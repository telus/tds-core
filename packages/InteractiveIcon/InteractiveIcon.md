### Minimal usage

`InteractiveIcon` are icons that have either motion/hover and/or active states to visually indicate that by interacting with it, a reaction will happen. Interactive icons must be accessible and announced by assistive technology.

```jsx
<Close a11yText="Close" />
```

### Accessibility

Each `InteractiveIcon` must have a minimum click/tap area of **40px** to ensure that they can be touched easily on mobile devices. Take care not to place other elements too close to an `InteractiveIcon`, as the touch area could overlap.

```jsx noeditor
const IconTable = require('../../docs/components/custom/IconTable/IconTable').default
;<>
  <IconTable
    heading="Icons"
    icons={[
      {
        name: 'Add',
        Component: Add,
        usageCriteria: 'Used to add items',
        props: {
          a11yText: 'Add',
        },
      },
      {
        name: 'Close',
        Component: Close,
        usageCriteria: 'Used to close',
        props: {
          a11yText: 'Close',
        },
      },
      {
        name: 'PlayVideo',
        Component: PlayVideo,
        usageCriteria:
          'Used to indicate that a video is available, used with text link to bring to another page or window',
        props: {
          a11yText: 'Play',
        },
      },
      {
        name: 'Subtract',
        Component: Subtract,
        usageCriteria: 'Used to reduce items',
        props: {
          a11yText: 'Remove',
        },
      },
    ]}
  />
</>
```

### Making InteractiveIcons interactive

Use the `onClick` property to add interactivity to an `InteractiveIcon`.

```js
initialState = {
  count: 0,
}

const addCount = () => {
  setState({ count: state.count + 1 })
}
const reduceCount = () => {
  setState({ count: state.count - 1 })
}
;<Box inline between={3} style={{ alignItems: 'center' }}>
  <Subtract onClick={reduceCount} a11yText="Reduce" />
  <Text size="medium">{state.count}</Text>
  <Add onClick={addCount} a11yText="Add" />
</Box>
```
