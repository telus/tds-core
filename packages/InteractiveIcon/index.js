export * from './svgs'
export { default as IconButton } from './IconButton'
