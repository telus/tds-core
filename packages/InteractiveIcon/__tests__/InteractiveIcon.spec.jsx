import React from 'react'
import { shallow, mount } from 'enzyme'

import InteractiveIcon, { StyledInteractiveIconButton } from '../InteractiveButton'

describe('InteractiveIcon', () => {
  const doShallow = (props = {}) =>
    shallow(
      <InteractiveIcon a11yText="This is an interactive icon" {...props}>
        <svg />
      </InteractiveIcon>
    )

  it('renders', () => {
    const interactiveIcon = mount(
      <InteractiveIcon a11yText="This is an interactive icon">
        <svg />
      </InteractiveIcon>
    )

    expect(interactiveIcon).toMatchSnapshot()
  })

  it('does other things', () => {
    const interactiveIcon = doShallow()

    expect(interactiveIcon).toExist()
  })

  it('passes additional attributes to the element', () => {
    const interactiveIcon = doShallow({ id: 'the-id', 'data-some-attr': 'some value' }).find(
      StyledInteractiveIconButton
    )

    expect(interactiveIcon).toHaveProp('id', 'the-id')
    expect(interactiveIcon).toHaveProp('data-some-attr', 'some value')
  })

  it('does not allow custom CSS', () => {
    const interactiveIcon = doShallow({
      className: 'my-custom-class',
      style: { color: 'hotpink' },
    })

    expect(interactiveIcon).not.toHaveProp('className', 'my-custom-class')
    expect(interactiveIcon).not.toHaveProp('style')
  })

  it('forwards refs', () => {
    const ref = React.createRef()
    const interactiveIcon = mount(
      <>
        <InteractiveIcon ref={ref} a11yText="This is an interactive icon">
          <svg />
        </InteractiveIcon>
      </>
    )
    const target = interactiveIcon
      .find('StyledComponent')
      .at(0)
      .childAt(0)
      .instance()

    expect(target).toEqual(ref.current)
  })
})
