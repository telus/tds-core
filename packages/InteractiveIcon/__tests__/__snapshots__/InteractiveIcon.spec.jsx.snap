// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InteractiveIcon renders 1`] = `
.c2 {
  position: absolute;
  height: 1px;
  width: 1px;
  overflow: hidden;
  -webkit-clip: rect(1px,1px,1px,1px);
  clip: rect(1px,1px,1px,1px);
}

.c1 {
  padding: 0;
  margin: 0;
  border-width: 0;
  font-family: "TELUS-Web","Helvetica Neue",Helvetica,Arial,sans-serif;
  color: #2a2c2e;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: none;
  box-shadow: none;
  cursor: pointer;
  width: 40px;
  height: 40px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  -webkit-tap-highlight-color: transparent;
}

.c1:focus {
  outline: none;
}

.c1:hover svg {
  -webkit-transform: scale(1.1,1.1);
  -ms-transform: scale(1.1,1.1);
  transform: scale(1.1,1.1);
}

.c1:active svg {
  -webkit-transform: scale(1,1);
  -ms-transform: scale(1,1);
  transform: scale(1,1);
}

.c3 {
  background-color: #d8d8d8;
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  z-index: 1;
  border-radius: 50%;
  -webkit-transition: -webkit-transform 200ms ease-in-out;
  -webkit-transition: transform 200ms ease-in-out;
  transition: transform 200ms ease-in-out;
  -webkit-transform: scale(0,0);
  -ms-transform: scale(0,0);
  transform: scale(0,0);
}

.c0:focus .c3,
.c0:active .c3 {
  -webkit-transform: scale(1,1);
  -ms-transform: scale(1,1);
  transform: scale(1,1);
}

@media (prefers-reduced-motion:reduce) {
  .c3 {
    -webkit-transition: none;
    transition: none;
  }
}

<button
  class="c0 c1"
>
  <span
    class="c2"
  >
    This is an interactive icon
  </span>
  <span
    class="c3"
  />
  <svg />
</button>
`;
